### PR TITLE
agent: add tests for do_write_stream function

### LINKED ISSUE #3984: agent tests: Add tests for do_write_stream function in src/rpc.rs
Add tests for `AgentService::do_write_stream` function in the `src/rpc.rs` file in the agent. Related to #292.

### PR DESCRIPTION
**Pull request by bradenrayhorn**
Add test coverage for do_write_stream function of AgentService
in src/rpc.rs. Includes minor refactoring to make function more
easily testable.

Fixes #3984

Signed-off-by: Braden Rayhorn <bradenrayhorn@fastmail.com>

### CODE REVIEW

File: src/agent/src/rpc.rs [hunk 1]
```diff
@@ -2016,6 +2055,140 @@ mod tests {
...
+                data: vec![],
+                result: Ok(WriteStreamResponse {
+                    len: 0,
+                    ..WriteStreamResponse::default()
+                }),
+                ..TestData::default()
+            },
+            TestData {
+                break_pipe: true,
+                result: Err(anyhow!("Broken pipe (os error 32)")),
```
Comment (bradenrayhorn) [inline src/agent/src/rpc.rs:2114]: One point I'd like feedback on is whether we should be asserting error messages that come from the operating system. It's something that might have the potential to change outside of the control of this repository.

File: src/agent/src/rpc.rs [hunk 2]
```diff
@@ -2016,6 +2055,140 @@ mod tests {
...
+                    &oci::Process::default(),
+                    &exec_process_id.to_string(),
+                    false,
+                    1,
+                )
+                .unwrap();
+
+                let fd = {
+                    if d.has_fd {
+                        let (rfd, wfd) = unistd::pipe().unwrap();
```
Comment (bradenrayhorn) [inline src/agent/src/rpc.rs:2151]: Also, is it worth adding code to close these file descriptors within the test? Or is it okay to let the OS handle these when the test runner finishes?

File: src/agent/src/rpc.rs (hunk 1 above)
Comment (fidencio) [reply src/agent/src/rpc.rs:2114]: I am not that comfortable doing this, but if we rely on that for something else already, then we better ensure a change on the message won't also break the "something else" relying on this.

File: src/agent/src/rpc.rs (hunk 2 above)
Comment (fidencio) [reply src/agent/src/rpc.rs:2151]: Please, close those within the test. :-).

File: src/agent/src/rpc.rs (hunk 1 above)
Comment (bradenrayhorn) [reply src/agent/src/rpc.rs]: I agree, I'm not too comfortable with it either. I've removed it for now. I'm open to ideas for how to test the error case of [this line](https://github.com/bradenrayhorn/kata-containers/blob/test-do-write-stream/src/agent/src/rpc.rs#L588). The ideas I've got seem like they'd complicate the test more than the value it'd provide.

Comment (jodh-intel) [reply src/agent/src/rpc.rs]: The best solution is to convert the `std::io::Error` to an O/S error (numeric `errno`) value in the test. You could do something like this in the test code (no error checking below but just shows the concept):

```rust
let result: std::io::Error = some_function();

// Convert the Error to a numeric value
let errno_value = e.raw_os_error().unwrap();

assert_eq!(errno_value, libc::EPIPE, "{}", msg);
```

Comment (bradenrayhorn) [reply src/agent/src/rpc.rs]: Thanks for the info @jodh-intel, I've updated the test to make its assertions in a similar method.

**Review by jodh-intel — COMMENTED** [review]
Thanks @bradenrayhorn.

I'm seeing `test_do_write_stream()` fail locally though:

```
$  export test=rpc::tests::test_do_write_stream; cargo test "$test" -vv -- --exact --nocapture
running 1 test
thread 'rpc::tests::test_do_write_stream' panicked at 'called `Result::unwrap()` on an `Err` value: EINVAL: Invalid argument', src/rpc.rs:1938:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test rpc::tests::test_do_write_stream ... FAILED

failures:

failures:
    rpc::tests::test_do_write_stream
```

**Comment by bradenrayhorn** [comment]
Thanks @jodh-intel, I'll be honest I'm a bit confused by the error you are seeing. In the commit you reviewed, [line 1938](https://github.com/kata-containers/kata-containers/commit/65ecfa342597cff9296d658b34c36afadebc34a2#diff-f87c1c212bba3abd0b07bafe69d5b01ad00dfe2c01bed585e2a75fcbb6442b89R1938) is not unwrapping anything, so I'm not sure where exactly this error is coming from.

**Review by jodh-intel — APPROVED** [review]
Thanks @bradenrayhorn.

lgtm

btw - The error I was seeing was caused by a bad install of rust fwics.

**Review by liubin — APPROVED** [review]

**Comment by bradenrayhorn** [comment]
I'm not sure why the metrics CI is failing. @jodh-intel or @liubin, if you have time, do you have any idea? Do we just need to retest again?

**Comment by liubin** [comment]
> I'm not sure why the metrics CI is failing. @jodh-intel or @liubin, if you have time, do you have any idea? Do we just need to retest again?

Don't worry about it, eventually, it will pass, we can retry it.

**Pull request merged**